File: CaseyCOMPILER/src/main/java/sharknoon/casey/compiler/java/generator/item/OnFunction.java
package sharknoon.casey.compiler.java.generator.item;

import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeSpec;
import sharknoon.casey.compiler.general.beans.CLIArgs;
import sharknoon.casey.compiler.general.beans.Item;
import sharknoon.casey.compiler.general.beans.Item.ItemType;

import javax.lang.model.element.Modifier;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

public class OnFunction {
    
    
    public static boolean accept(CLIArgs args, Path currentPath, Item item) {
        Optional<MethodSpec> optionalFunction = ItemUtils.getFunction(args, item, true);
        if (!optionalFunction.isPresent()) {
            System.err.println("Could not create static function " + ItemUtils.getFullName(item));
            return false;
        }
        String className = item.name;
        List<MethodSpec> methods = new ArrayList<>();
        MethodSpec function = optionalFunction.get();
        methods.add(function);
        
        if (ItemUtils.isMainMethod(args, item)) {
            Map<String, String> parameters = args.getParameters();
            List<String> paramterValuesInRightOrder = item.children.stream()
                    .filter(i -> i.item == ItemType.PARAMETER)
                    .map(n -> {
                        if (parameters.containsKey(n.name)) {
                            if ("TEXT".equals(n.type)) {
                                return "\"" + parameters.get(n.name) + "\"";
                            }
                            return parameters.get(n.name);
                        } else {
                            System.err.println("Function "
                                    + ItemUtils.getFullName(item)
                                    + " has wrong parameters actual: ("
                                    + String.join(", ", parameters.keySet())
                                    + ") expected: ("
                                    + item.children.stream()
                                    .filter(p -> p.item == ItemType.PARAMETER)
                                    .map(p -> p.name)
                                    .collect(Collectors.joining(", "))
                                    + ")"
                            );
                            return null;
                        }
                    })
                    .collect(Collectors.toList());
            
            if (paramterValuesInRightOrder.contains(null)) {
                return false;
            }
            
            MethodSpec mainMethod = MethodSpec.methodBuilder("main")
                    .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
                    .addParameter(String[].class, "args")
                    .addStatement(className + "." + className + "(" + String.join(", ", paramterValuesInRightOrder) + ")")
                    .build();
            methods.add(mainMethod);
        }
        
        try {
            TypeSpec clazz = TypeSpec.classBuilder(className)
                    .addModifiers(Modifier.PUBLIC)
                    .addMethods(methods)
                    .build();
            JavaFile varFile = JavaFile.builder(ItemUtils.pathToClassPath(currentPath), clazz)
                    .build();
            varFile.writeTo(args.getBasePath());
        } catch (Exception e) {
            System.err.println("Could not create function " + ItemUtils.getFullName(item) + ": " + e);
            return false;
        }
        return true;
    }
    
}

File: CaseyCOMPILER/src/main/java/sharknoon/casey/compiler/java/generator/item/OnFunction.kt
package sharknoon.casey.compiler.java.generator.item

import com.squareup.javapoet.*
import sharknoon.casey.compiler.general.beans.*
import sharknoon.casey.compiler.general.beans.Item.ItemType
import java.nio.file.Path
import java.util.stream.*
import javax.lang.model.element.Modifier.*


fun acceptFunction(args: CLIArgs, currentPath: Path, item: Item): Boolean {
    val function = getFunction(args, item, true)
    if (function == null) {
        System.err.println("Could not create static function " + getFullName(item))
        return false
    }
    val className = item.name
    val methods = mutableListOf<MethodSpec>()
    methods.add(function)

    if (isMainMethod(args, item)) {
        val parameters = args.parameters
        val parameterValuesInRightOrder = item.children.stream()
                .filter { i -> i.item === ItemType.PARAMETER }
                .map { n ->
                    if (parameters.containsKey(n.name)) {
                        if ("TEXT" == n.type) {
                            return@map "\"" + parameters[n.name] + "\""
                        }
                        return@map parameters[n.name]
                    } else {
                        System.err.println("Function "
                                + getFullName(item)
                                + " has wrong parameters actual: ("
                                + parameters.keys.joinToString(", ")
                                + ") expected: ("
                                + item.children.stream()
                                .filter { p -> p.item === ItemType.PARAMETER }
                                .map { p -> p.name }
                                .collect(Collectors.joining(", "))
                                + ")"
                        )
                        return@map null
                    }
                }
                .collect(Collectors.toList())

        if (parameterValuesInRightOrder.contains(null)) {
            return false
        }

        val mainMethod = MethodSpec.methodBuilder("main")
                .addModifiers(PUBLIC, STATIC)
                .addParameter(Array<String>::class.java, "args")
                .addStatement(className + "." + className + "(" + parameterValuesInRightOrder.joinToString(", ") + ")")
                .build()
        methods.add(mainMethod)
    }

    try {
        val clazz = TypeSpec.classBuilder(className)
                .addModifiers(PUBLIC)
                .addMethods(methods)
                .build()
        val varFile = JavaFile.builder(pathToClassPath(currentPath), clazz)
                .build()
        varFile.writeTo(args.basePath)
    } catch (e: Exception) {
        System.err.println("Could not create function " + getFullName(item) + ": " + e)
        return false
    }

    return true
}


